packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/CollectionResultDefinition.cs

```diff
@@ -0,0 +1,324 @@
+using Microsoft.TypeSpec.Generator.ClientModel.Snippets;
+using Microsoft.TypeSpec.Generator.Input;
+using Microsoft.TypeSpec.Generator.Primitives;
+using Microsoft.TypeSpec.Generator.Providers;
+using Microsoft.TypeSpec.Generator.Snippets;
+using Microsoft.TypeSpec.Generator.Statements;
+using static Microsoft.TypeSpec.Generator.Snippets.Snippet;
+
+namespace Microsoft.TypeSpec.Generator.ClientModel.Providers
+{
+    internal class CollectionResultDefinition : TypeProvider
+    {
+        private readonly InputOperation _operation;
+        private readonly bool _isAsync;
+        private readonly ClientProvider _client;
+        private FieldProvider _clientField;
+        private readonly FieldProvider _optionsField;
+        private readonly CSharpType _responseType;
+        private readonly CSharpType? _modelType;
+        private readonly string? _nextLinkPropertyName;
+        private readonly FieldProvider _initialUri;
+        private readonly InputResponseLocation? _nextLinkLocation;
+
+        private static ParameterProvider PageParameter =
+            new("page", FormattableStringHelpers.Empty, new CSharpType(typeof(ClientResult)));
+
+        private readonly string _itemsPropertyName;
+
+        public CollectionResultDefinition(ClientProvider client, InputOperation operation, CSharpType? modelType, bool isAsync)
+        {
+            _client = client;
+            _clientField = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                _client.Type,
+                "_client",
+                this);
+            _initialUri = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                new CSharpType(typeof(Uri)),
+                "_initialUri",
+                this);
+            _optionsField = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                new CSharpType(typeof(RequestOptions)),
+                "_options",
+                this);
+
+            _modelType = modelType;
+            _operation = operation;
+            _isAsync = isAsync;
+
+            var response = _operation.Responses.FirstOrDefault(r => !r.IsErrorResponse);
+            var model = ScmCodeModelPlugin.Instance.TypeFactory.CreateModel((InputModelType)response!.BodyType!)!;
+            // TODO Nested models are not supported yet https://github.com/Azure/typespec-azure/issues/2287
+            var nextLinkPropertyName = _operation.Paging!.NextLink?.ResponseSegments[0];
+            _nextLinkLocation = operation.Paging!.NextLink?.ResponseLocation;
+
+            // TODO Nested models are not supported https://github.com/Azure/typespec-azure/issues/2287
+            var itemPropertyName = _operation.Paging!.ItemPropertySegments[0];
+            _itemsPropertyName = model.CanonicalView.Properties
+                .FirstOrDefault(p => p.WireInfo?.SerializedName == itemPropertyName)?.Name ?? itemPropertyName;
```

Comment:
I think we should report something when we did not find the itemPropertyName and has to fallback - this has very high probability to generate code that could not compile.